packages/circuits/test/proveGraffitiPreimage.test.ts

```diff
@@ -0,0 +1,154 @@
+    const commitment = id.genIdentityCommitment()
+    const hashedLeaf = crypto.hashLeftRight(commitment, userStateRoot)
+    GSTree.insert(hashedLeaf)
+    const GSTreeProof = GSTree.createProof(0) // if there is only one GST leaf, the index is 0
+    const GSTreeRoot = GSTree.root
+
+    const circuitInputs = {
+        epoch,
+        identity_nullifier: id.identityNullifier,
+        identity_trapdoor: id.trapdoor,
+        user_tree_root: userStateRoot,
+        GST_path_index: GSTreeProof.pathIndices,
+        GST_path_elements: GSTreeProof.siblings,
+        attester_id: attesterId,
+        pos_rep: reputationRecords[attesterId].posRep,
+        neg_rep: reputationRecords[attesterId].negRep,
+        graffiti: reputationRecords[attesterId].graffiti,
+        sign_up: reputationRecords[attesterId].signUp,
+        UST_path_elements: USTPathElements,
+        graffiti_pre_image: graffitiPreImage,
+    }
+    return crypto.stringifyBigInts(circuitInputs)
+}
+
+const genProofAndPublicSignals = async (inputs: any): Promise<any> => {
+    const circuitWasmPath = path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.wasm`
+    )
+    const zkeyPath = path.join(__dirname, `../zksnarkBuild/${circuitName}.zkey`)
+    const { proof, publicSignals } = await snarkjs.groth16.fullProve(
+        inputs,
+        circuitWasmPath,
+        zkeyPath
+    )
+    return { proof, publicSignals }
+}
+
+const verifyProof = async (publicSignals, proof): Promise<boolean> => {
+    const vkey = require(path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.vkey.json`
+    ))
+    return snarkjs.groth16.verify(vkey, publicSignals, proof)
+}
+
+describe('Prove grafitti preimage', function () {
+    this.timeout(300000)
+
+    const epoch = 1
+    const nonce = 1
+
+    it('successfully prove graffiti preimage', async () => {
+        const attesterId = 1
+        const reputationRecords = {
+            [attesterId]: new Reputation(
+                BigInt(10),
+                BigInt(0),
+                hashOne(BigInt(graffitiPreImage)),
+                BigInt(1)
+            ),
+        }
+        const circuitInputs = genCircuitInput(
+            new ZkIdentity(),
+            epoch,
+            reputationRecords,
+            attesterId
+        )
+
+        const { proof, publicSignals } = await genProofAndPublicSignals(
+            circuitInputs
+        )
+        const isValid = await verifyProof(publicSignals, proof)
+        expect(isValid).to.be.true
+    })
+
+    it('should fail to prove a reputation with wrong graffiti preimage', async () => {
+        const attesterId = 1
+        const reputationRecords = {
+            [attesterId]: new Reputation(
+                BigInt(10),
+                BigInt(0),
+                hashOne(BigInt(graffitiPreImage)),
+                BigInt(1)
+            ),
+        }
+        const id = new ZkIdentity()
+        const wrongGraffitiPreImage = graffitiPreImage + 1
```

Comment:
I think this works, but you should also test with a non-zero graffiti pre-image. This is why you should pass it as a variable to `genCircuitInput` so the function can calculate the currect user/state tree leaf.

Reply:
so should I make another argument in`genCircuitInput` for preimages and should I pass `wrongGraffitiPreimage` there for the failing test case?

Reply:
I don't know if I interpreted comments from you guys correctly, but I made a modification about this. I also added another test with non-zero graffiti preimage. but these tests are not passing any more. any idea why??

Reply:
@vimwitch Sorry to bother you, but I'd appreciate it if you can comment on this!

Reply:
Yep, i think it looks good. I would remove the top level `graffitiPreImage` variable and instead define it for each test function